modules/engine/src/model/model.ts

```diff
@@ -291,9 +289,33 @@ export class Model {
    * @note Triggers a pipeline rebuild / pipeline cache fetch on WebGPU
    */
   setBufferLayout(bufferLayout: BufferLayout[]): void {
-    if (bufferLayout !== this.bufferLayout) {
-      this.bufferLayout = bufferLayout;
-      this._setPipelineNeedsUpdate('bufferLayout');
+    this.bufferLayout = this._gpuGeometry
+      ? mergeBufferLayouts(bufferLayout, this._gpuGeometry.bufferLayout)
+      : bufferLayout;
+    this._setPipelineNeedsUpdate('bufferLayout');
+
+    // Recreate the pipeline
+    this.pipeline = this._updatePipeline();
+
+    const oldVertexArray = this.vertexArray;
+    // vertex array needs to be updated if we update buffer layout,
+    // but not if we update parameters
+    this.vertexArray = this.device.createVertexArray({
+      renderPipeline: this.pipeline
+    });
+
+    // Transfer previously set attributes
+    if (oldVertexArray) {
```

Comment:
Would it be reasonable to ask the app to repopulate its vertexArray if it changes bufferLayout?

I am saying that because this auto copying of attributes looks a bit dicey. I have a feeling that changing the buffer layout could seriously invalidate those old attributes and leave things in hard to debug states...

The exception would perhaps be the geometry attributes which we could set (not copy) at this stage.

Reply:
These can be repopulated on the app side. I put the code here primarily because the geometry attributes are lost when vertex array is recreated.